### ISSUE TITLE
easyeffects consumes 3-5 % fo CPU even if mic is not used

### ISSUE THREAD

**Issue by promeneur** [wwmm/easyeffects#1704]
### EasyEffects Version

6.2.6

### What package are you using?

openSUSE

### Distribution

openSUSE Tumbleweed

### Describe the bug

I set easyeffects according to cancel noise for the mic.
I check that easyeffects consumes 3-5 % CPU when I don't use the mic.

Furthermore, I was thinking that easyeffetcs is in some sort of sleep mode when we don't use the mic.

Noisetorch doesn't use any % CPU when the mic is not used.

### Expected Behavior

easyeefects doesn't use any % CPU when the mic is not used.

### Debug Log

<details><summary>Debug Log</summary><br/>

```
Paste your log here
```
</details>

**Comment by wwmm** [comment]
I can't reproduce this issue. When nothin is recording from our virtual source our pipeline is unlinked and as a result no processing is done. What may be happening to you is that thirds party tools as Pavucontrol or anything else that can be attaching level meters tou our virtual source are forcing our pipeline to stay active.

This happens because at the end of the day level meters links looks like any other link. So if you keep Pavucontrol or similar programs opened while easyeffects is running they will force our pipeline to stay linked.

**Comment by wwmm** [comment]
Something else to keep in mind is that EasyEffects has two pipelines. One for the mic and another for the speakers. Are you sure there is no activity on the speakers pipeline?

**Comment by promeneur** [comment]
>>  What may be happening to you is that third party tools as Pavucontrol or anything else that can be attaching level meters to >> our virtual source are forcing our pipeline to stay active.

I don't use pavucontrol

**Comment by wwmm** [comment]
>I don't use pavucontrol

Ok. But there is something besides your audio recording application linked to our virtual device? You can see this through Helvum or by running `pw-dot`. It's output file can be viewed with xdot. 

Something else to have in mind is that there is an activity timeout we use in order to decide if we have to unlink the filters or not. You can customize it in oir settings menu.

**Comment by promeneur** [comment]
I noticed this:

without easyeffects 

if I speak, then I see the mic volume changing  and the speakers volume stays at 0.

with easyeffects
if i speak then i see the mic volume changing and the speakers volume changing.

Note: I disabled easyeffects for all outputs because I just want only noise cancelation for the mic.

**Comment by promeneur** [comment]
>> Ok. But there is something besides your audio recording application linked to our virtual device? You can see this through 
>> Helvum 

I attached several captures of my settings. One of them is about helvum

**Comment by promeneur** [comment]
When measuring % CPU I don't use any app as Skype for example. then no app uses the mic

**Comment by wwmm** [comment]
>
I attached several captures of my settings. One of them is about helvum

I do not see any attached image in this issue. But I am on my phone at this moment. Maybe the mobile version of this site is hiding something.

**Comment by promeneur** [comment]
Sorry the attachments are here #1703 

a new helvum view of pipewire 

![pipewire_helvum](https://user-images.githubusercontent.com/240424/181776556-adc8d0de-7302-478f-8ca6-704c81502127.png)

**Comment by wwmm** [comment]
>Sorry the attachments are here https://github.com/wwmm/easyeffects/issues/1703

Oh... Ok. Is that Plas stream visible all the time? It will have the same effect as Pavucontrol. I thought I had already Blocklisted plasma. Is there any change if you add it's stream to the user Blocklist?

**Comment by promeneur** [comment]
>> you add it's stream to the user Blocklist?

How to ?

**Comment by promeneur** [comment]
I made an experiment.

I shut down a QEMU VM using mic and speakers of the host.

Result : easyeffects consumes 0 % CPU

Note: no app in the VM using mic and speakers is running.

**Comment by wwmm** [comment]
>How to ?

At the side of the enable checkbox there is one called `Exclude`. Enable it.

But thinking about it again it does not seem it is going to be enough. I will have to find a way to deal with these level meter streams once and for all. Differently from Pavucontrol I do not think you can "just close" this one from plasma...

**Comment by wwmm** [comment]
>I shut down a QEMU VM using mic and speakers of the host.

Oh... Só qemu has streams that are active all the time... This explains why our pipeline is always linked in your system. Try to enable the exclude checkbox in it's stream

**Comment by promeneur** [comment]
Another result of my experiment : in the easyeffects setting tool now I get an empty list for input and ouput players

**Comment by promeneur** [comment]
Another result of my experiment 

![pipewire_helvum_1](https://user-images.githubusercontent.com/240424/181781699-4e4f2dfa-2df5-4bbc-80b7-f110db43a4f7.png)

**Comment by promeneur** [comment]
I excluded the input and output of the QEMU VM.

Resutl:  easyeffects consumes 0 % CPU.

But now there is no noise cancelation for any app running in the VM. No ?

**Comment by promeneur** [comment]
In plasma volume control, we can see that QEMU VM mic and speakers are disabled.

![plasma_volume_control](https://user-images.githubusercontent.com/240424/181788016-5ddfb379-c1d0-4c4a-a228-3a2d3a4c0ef3.png)

**Comment by promeneur** [comment]
Can I conclude that easyeffects does not take care of this ?

Why for easyeffects they are enabled ?

In QEMU VM mic and speakers are enabled on demand. For example, if a notification happens or I use an app using the mic.

**Comment by wwmm** [comment]
>In plasma volume control, we can see that QEMU VM mic and speakers are disabled.

Unfortunately what that Plasma window is telling is that these streams are muted. A muted stream can still have active links. And what easyeffects does in order to decide if our filters should stay linked is checking if there are active links linked to our virtual device. QEMU links are muted but probably still in the active state.

**Comment by wwmm** [comment]
Run `pe-dump` while the QEMU streams are muted and linked to easyeffects. It's output file should show in what state the QEMU links really are. I think I will have to add some kind of workaround for it. I am not sure what yet.

**Comment by promeneur** [comment]
pw-dump log

[pw-dup.log](https://github.com/wwmm/easyeffects/files/9221238/pw-dup.log)

**Comment by wwmm** [comment]
>[pw-dup.log](https://github.com/wwmm/easyeffects/files/9221238/pw-dup.log)

In this file the node `easyeffects_source` has `id = 37` and can be seen at the line `1210`. The `QEMU` node has `id = 128` and can be seen at the line `9019`. The state of one of the links between these two nodes can be seen at the line `9879`. As expected the link is in the active state. Thinking about it `pw-dot` would have been a more straightforward way to find this information as it shows the link state too :smile: .

In any case now I understand why our pipeline stays active. The question is if there is something that can be done about it. Does QEMU really need to keep its streams in the active state even if no application is actually trying to use its virtual microphone? I think this is something we have to find out somehow. As far as the `pw-dump` output goes I do not see anything special about the QEMU stream properties I could use to somehow allow EE to unlink its filters even if the QEMU stream is active. 

Ideally QEMU should `pause` its stream when nothing is recording from its virtual devices. But for some reason it does not do that. Muting the stream is not enough. It has to be paused or stopped.

**Comment by promeneur** [comment]
I use these statements for the audio part of the QEMU VM

-device intel-hda \
-device hda-duplex,audiodev=snd0 \
-audiodev pa,id=snd0 \

This creates an Intel audio HD  sound card snd0 with the driver pa (pulseaudio) of the host.

**Comment by samuk190** [comment]
![image](https://user-images.githubusercontent.com/19878001/182051061-00d008e0-2ce2-49e2-a930-50c59c1cc45a.png)

it must be a bug from your platform, Im using ryzen and I have 1% usage on idle while listening music, weird...

**Comment by wwmm** [comment]
>it must be a bug from your platform, Im using ryzen and I have 1% usage on idle while listening music with my focusrite 2i2, weird...

The amount of CPU usage can very a lot of one model to another. The real problem here is QEMU keeping its streams in the active state all the time. That is why EasyEffects keeps using CPU even when there is not application in the virtual image actually using the QEMU virtual device.

**Comment by promeneur** [comment]
Perhaps it is not QEMU the problem, but the host OS : android-x86 (android-x86.org).

I attach  the scripts to create, install, execute android 7.1-r5 from android-x86.org if you want to test.

https://sourceforge.net/projects/android-x86/files/Release%207.1/android-x86_64-7.1-r5.iso/download

[android_x86_7.1-r5_disk_create.sh.txt](https://github.com/wwmm/easyeffects/files/9232084/android_x86_7.1-r5_disk_create.sh.txt)

[android_x86_7.1-r5_install.sh.txt](https://github.com/wwmm/easyeffects/files/9232090/android_x86_7.1-r5_install.sh.txt)

[android_x86_7.1-r5_start.sh.txt](https://github.com/wwmm/easyeffects/files/9232146/android_x86_7.1-r5_start.sh.txt)

**Comment by promeneur** [comment]
My CPU is an Intel i5-7400

**Comment by wwmm** [comment]
I still would not blame the CPU. When the condition to unlink our filters is met you will see zero percent CPU usage in any CPU. But the way qemu is handling it's streams is not allowing this to happen.

**Comment by wwmm** [comment]
What qemu is doing is the same as having an audio recorder continually running on the background. For EE it is like any other audio recorder stream and as a result our filters stay linked and active. QEMU should put it's stream in the active state only when it is needed and pause or destroy when not needed.

**Comment by promeneur** [comment]
AS I said in the description of the problem : Noisetorch doesn't use any % CPU power.

How do you explain this ?

**Comment by promeneur** [comment]
Perhaps noisetorch is activating when hearing a voice. No ?

**Comment by wwmm** [comment]
>Perhaps noisetorch is activating when hearing a voice. No ?

This possibility crossed my mind. But in other to detect voice you have to be recording from the microphone. What would result in CPU being used. So I imagine noisetorch is doing something else in order to decide when to unlink its filters. Maybe they are creating filters using a different pipewire object. Pipewire offers more than one way to create filters.

**Comment by wwmm** [comment]
Is noisetorch using pipewire's API or does it still uses pulse audio code? I searched for the words `pw_filter` and `pw_stream`  in the noisetorch-ng repository and there was no return.

If noisetorch is still using pulseaudio code it definitely is doing things in a very different way. And we may actually be seeing some kind of pipewire limitation. Either in easyeffects or in QEMU.

**Comment by wwmm** [comment]
>If noisetorch is still using pulseaudio code it definitely is doing things in a very different way.

I took a look at its code now and that is what happening. Noisetorch doesn't have to handle a pipeline. They just load a virtual device and some other modules. Things are way more complicated when doing what EasyEffects has to do.

**Comment by promeneur** [comment]
OK Thanks

**Comment by karolherbst** [comment]
I've noticed this issue also happens with audio players like spotify or simply just firefox. Using the Bass Enhancer and Crystalizer effects. Version 7.2.3. Should there be a new issue for outputs causing constant CPU load or is this fine to keep it here?

Normally this issue only starts appearing after those applications did play something in the past but are now paused or not causing any audio output.

**Comment by wwmm** [comment]
>Should there be a new issue for outputs causing constant CPU load or is this fine to keep it here?

If there is I do not notice on my computer. Try to run `pw-top` in a situation you consider no CPU should be used. It will show i the plugins are active. It may also be interesting to see the file generated by `pw-dot` under the same circumstances. It can be viewed with the commad xdot and will show the links between filters and devices as well as their state.

**Comment by karolherbst** [comment]
```
S   ID  QUANT   RATE    WAIT    BUSY   W/Q   B/Q  ERR FORMAT           NAME 
S   29      0      0    ---     ---   ---   ---     0                  Dummy-Driver
S   30      0      0    ---     ---   ---   ---     0                  Freewheel-Driver
S   51      0      0    ---     ---   ---   ---     0                  Midi-Bridge
S   54      0      0    ---     ---   ---   ---     0                  bluez_midi.server
R   58    512  48000 663,4us   0,5us  0,06  0,00  602    S32LE 2 48000 alsa_input.usb-BEHRINGER_UMC202HD_192k-00.pro-input-0
R   57      0      0   7,1us   4,4us  0,00  0,00    6    S32LE 2 48000  + alsa_output.usb-BEHRINGER_UMC202HD_192k-00.pro-output-0
R   77    900  48000  79,3us  12,4us  0,01  0,00   28    F32LE 2 48000  + Firefox
R   72    900  48000  92,1us   3,4us  0,01  0,00   28    F32LE 2 48000  + Firefox
R   87      0      0   6,1us   6,2us  0,00  0,00    0     F32P 2 48000  + easyeffects_sink
R   97      0      0   1,4us   5,2us  0,00  0,00    1                   + ee_soe_output_level
R  102      0      0   2,5us   2,0us  0,00  0,00   10                   + ee_soe_spectrum
R  113      0      0  11,8us  92,1us  0,00  0,01  265                   + ee_soe_bass_enhancer
R  115      0      0   2,3us 425,6us  0,00  0,04  102                   + ee_soe_crystalizer
S   78      0      0    ---     ---   ---   ---     0                  v4l2_input.pci-0000_00_14.0-usb-0_9_1.0
S   80      0      0    ---     ---   ---   ---     0                  v4l2_input.pci-0000_00_14.0-usb-0_9_1.2
S   88      0      0    ---     ---   ---   ---     0                  easyeffects_source
I  127      0      0   0,0us   0,0us  ???   ???     0                  ee_sie_output_level
I  132      0      0   0,0us   0,0us  ???   ???     0                  ee_sie_spectrum
I  152      0      0   0,0us   0,0us  ???   ???     0    F32LE 2 44100 spotify
```

and pw-dot just shows spotify and firefox linked to the effects, but neither is actually outputting any sound.

**Comment by karolherbst** [comment]
anyway, doesn't seem like spotify is causing the load, because just quitting Firefox makes the CPU load go back to 0

**Comment by karolherbst** [comment]
> > and pw-dot just shows spotify and firefox linked to the effects, but neither is actually outputting any sound.
> 
> Hum... But based on the pw-top output they are sending audio buffers to EasyEffects. What explain why EasyEffects is still using CPU power. It seems Firefox is playing silence for some reason.

yeah.. probably something like that, still annoying, and not sure if it's firefox being misconfigured or just firefox doing that without reason.

**Comment by wwmm** [comment]
>yeah.. probably something like that, still annoying, and not sure if it's firefox being misconfigured or just firefox doing that without reason.

My guess is that a webpage you've accessed is asking it to activate a playback stream. I use Firefox regularly and I haven't noticed this strange behavior. So it does not seem to be something it does on its own.

**Comment by karolherbst** [comment]
yeah.. I just found this firefox extensions that helps to resolve this issue.. https://addons.mozilla.org/en-US/firefox/addon/audiocontext-suspender/

**Comment by wwmm** [comment]
So much has changed that is probably better to close this issue
